### ISSUE TITLE
Support path-style addressing

### ISSUE THREAD

**Issue by AndreyBozhko** [mygrocerydeals/gradle-s3-plugin#20]
It would be helpful to have the plugin support path-style addressing for S3 buckets.

More details about path-style addressing here:
- https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#path-style-access
- https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/examples-s3.html

**Comment by peter-thomas-mgd** [comment]
@AndreyBozhko path-style s3 bucket names are deprecated. How would the plugin benefit from introducing support for this naming convention and what use cases would it apply to?

**Comment by AndreyBozhko** [comment]
@peter-thomas-mgd This would allow the plugin to be used with proprietary S3 providers where path-style addressing is the only option.

**Comment by AndreyBozhko** [comment]
Or perhaps it makes more sense to add a generic hook so that users could customize the S3 client builder?

**Comment by AndreyBozhko** [comment]
Hi @peter-thomas-mgd, did you have a chance to look at this? Thanks

**Comment by peter-thomas-mgd** [comment]
`v1.4.0` introduced support for custom endpoints, which already customizes the S3 client for an alternate S3 providers. It sounds like additional options are needed for the provider you are working with. Based on the current implementation of the plugin, a number of scenarios that would likely break if we introduced an alternate way to specify bucket names and path expressions (for example, wildcard matching for downloads).
Can you provide a complete example (e.g. single file upload/download, directory upload/download, S3 provider) with a sample URL that requires support?

**Comment by AndreyBozhko** [comment]
Something like this works for my use case:

```kotlin
open class PatchedS3Download : S3Download() {
    override fun getAsyncS3Client(): S3AsyncClient {
        return S3AsyncClient.crtBuilder()
            .credentialsProvider(credentialsChain)
            .endpointOverride(parsedEndpoint)
            .region(parsedRegion)
            .forcePathStyle(true)  // <-- pass one more configuration to S3Client
            .build()
    }
}
```

This seems like a straightforward change, and I'm not sure why that should break any plugin logic - it's one more configuration passed to the underlying S3Client. From what I can see, this configuration is not deprecated in the client SDK.

**Comment by AndreyBozhko** [comment]
It's possible to test the behavior using LocalStack (https://docs.localstack.cloud/getting-started/installation/) to emulate the S3 service locally:
1. Install LocalStack CLI and start the service.
2. Follow the steps in https://docs.localstack.cloud/user-guide/aws/s3/ to create a sample bucket and add a file to the bucket.
3. Verify that all looks good with `aws --endpoint-url=http://localhost:4566 s3 ls s3://sample-bucket/`
4. For each of the endpoint urls below, run the S3Download task to download the file from the emulated bucket.

```kotlin
tasks.register<S3Download>("s3-test") {
    endpoint = "http://localhost:4566"  // or http://s3.localhost.localstack.cloud:4566
    region = "us-east-1"
    bucket = "sample-bucket"
    key = "sample-file"
    file = "path/where/to/save"

    doNotTrackState("always re-run")
}
```

When the service is running, any of the endpoints `localhost:4566` or `s3.localhost.localstack.cloud:4566` can be used to interact with the service.

However, the tool documentation indicates (https://docs.localstack.cloud/user-guide/aws/s3/#path-style-and-virtual-hosted-style-requests) that LocalStack will only accept path-style requests to the first endpoint, and will error our if the incoming request uses virtual hosted-style request.

So the task above fails with the first endpoint, and succeeds with the second endpoint.

But if the appropriate `forcePathStyle` configuration is passed to the S3Client, the task succeeds for the `localhost:4566` endpoint as well.

**Comment by peter-thomas-mgd** [comment]
Thanks for the additional details. I understand the use case now. I agree that it is straightforward to add an additional optional parameter for custom endpoints.

So now back to the original question: why should we add support for a deprecated, non-recommended URI style that will only apply to custom endpoints? Both Amazon and LocalStack recommend virtual-style URIs. It seems easy enough to update the Gradle build to follow virtual-style endpoint URIs in the task configuration. The whole point of the Gradle plugin is so that people don't have to deal with these kinds of semantics.

**Comment by AndreyBozhko** [comment]
I agree that for interacting with AWS S3, the path-style configuration isn't really relevant. Although various open source libraries/tools for interacting with S3 ([aws-java-nio-spi-for-s3](https://github.com/awslabs/aws-java-nio-spi-for-s3), [mountpoint-s3](https://github.com/awslabs/mountpoint-s3) typically allow users to configure addressing style).

That being said, the plugin does offer support for other S3-compatible object storage providers besides AWS S3. So if the provider doesn't support virtual-hosted style addressing (e.g., Oracle's Object Storage), people would have to deal with the path-style semantics anyway.

So it's mostly the matter of convenience - to be able to set the path style configuration together with other configurations (if needed), and not have to figure out which all parts of the plugin need to be overridden.

**Comment by peter-thomas-mgd** [comment]
Appreciate the thoughtful feedback. I'll work towards adding support for this option and updating the docs. Hopefully I'll something ready in the next few days.

**Comment by peter-thomas-mgd** [comment]
Support for this capability has been added in `v2.1.0`.

**Comment by AndreyBozhko** [comment]
Thanks, really appreciate the quick turnaround!
